src/Pos/Block/Logic/Creation.hs

```diff
@@ -252,30 +279,57 @@ createMainBlockPure
        (MonadError Text m, SscHelpersClass ssc)
     => Byte                   -- ^ Block size limit (real max.value)
     -> BlockHeader ssc
-    -> [TxAux]
     -> Maybe ProxySKEither
     -> SlotId
-    -> [ProxySKHeavy]
-    -> SscPayload ssc
-    -> UpdatePayload
     -> SecretKey
+    -> RawPayload ssc
     -> m (MainBlock ssc)
-createMainBlockPure limit prevHeader txs pSk sId psks sscData usPayload sk =
-    flip evalStateT limit $ do
-        -- default ssc to put in case we won't fit a normal one
-        let defSsc = (sscDefaultPayload @ssc (siSlot sId) :: SscPayload ssc)
-
+createMainBlockPure limit prevHeader pSk sId sk rawPayload = do
+    bodyLimit <- execStateT computeBodyLimit limit
+    body <- createMainBody bodyLimit prevHeader pSk sId sk rawPayload
+    mkMainBlock (Just prevHeader) sId sk pSk body extraH extraB
+  where
+    extraB :: MainExtraBodyData
+    extraB = MainExtraBodyData (mkAttributes ())
+    extraH :: MainExtraHeaderData
+    extraH =
+        MainExtraHeaderData
+            lastKnownBlockVersion
+            curSoftwareVersion
+            (mkAttributes ())
+    -- default ssc to put in case we won't fit a normal one
+    defSsc :: SscPayload ssc
+    defSsc = sscDefaultPayload @ssc (siSlot sId)
+    computeBodyLimit :: StateT Byte m ()
+    computeBodyLimit = do
         -- account for block header and serialization overhead, etc;
         let musthaveBody = BC.MainBody
-                (fromMaybe (error "createMainBlockPure: impossible") $ mkTxPayload mempty)
-                defSsc [] def
+                (leftToPanic @Text "createMainBlockPure: impossible " $
+                 mkTxPayload mempty)
+                defSsc def def
         musthaveBlock <-
             mkMainBlock (Just prevHeader) sId sk pSk musthaveBody extraH extraB
         let mhbSize = biSize musthaveBlock
         when (mhbSize > limit) $ throwError $
             "Musthave block size is more than limit: " <> show mhbSize
         identity -= biSize musthaveBlock
 
+-- Main purpose of this function is to create main block's body taking
+-- limit into account. Usually this function doesn't fail, but we
+-- perform some sanity checks just in case.
+--
+-- Given limit applies only to body, not to other data from block.
+createMainBody
+    :: forall m ssc .
+       (MonadError Text m, SscHelpersClass ssc)
+    => Byte  -- ^ Body limit
+    -> SlotId
+    -> RawPayload ssc
+    -> m $ Body $ MainBlockchain ssc
+createMainBody bodyLimit sId payload =
```

Comment:
I don't know where it's more complicated, it does part of what it was doing before. Also it unwraps `DlgPayload` and `RawPayload`, but that's because, first, previously there was no `DlgPayload`, second, I added `RawPayload` to avoid passing 4 arguments to a function which already has too many arguments.